pints/_mcmc/_monomial_gamma_hamiltonian.py

```diff
@@ -0,0 +1,519 @@
+        return f
+
+    def _K(self, v_p, a, c, m):
+        """
+        Soft kinetic energy function defined in [1].
+        """
+        return np.sum([self._K_indiv(p, a, c, m) for p in v_p])
+
+    def _K_deriv(self, v_p, a, c, m):
+        """
+        Derivative of soft kinetic energy function defined in [1].
+        """
+        return np.array([self._K_deriv_indiv(p, a, c, m) for p in v_p])
+
+    def _K_deriv_indiv(self, p, a, c, m):
+        """
+        Derivative of soft kinetic energy function defined in [1]
+        for individual momentum component.
+        """
+        abs_p = np.abs(p)
+        sign_p = np.sign(p)
+        tanh = np.tanh(0.5 * c * abs_p**(1.0 / a) * sign_p / m)
+        return abs_p**(-2 + 1.0 / a) * p * sign_p * tanh / (a * m)
+
+    def _K_indiv(self, p, a, c, m):
+        """
+        Soft kinetic energy function defined in [1]
+        for individual momentum component.
+        """
+        return -self._g(p, a, m) + (2.0 / c) * np.log(
+            1.0 + np.exp(c * self._g(p, a, m)))
+
+    def leapfrog_steps(self):
+        """
+        Returns the number of leapfrog steps to carry out for each iteration.
+        """
+        return self._n_frog_iterations
+
+    def leapfrog_step_size(self):
+        """
+        Returns the step size for the leapfrog algorithm.
+        """
+        return self._step_size
+
+    def _log_init(self, logger):
+        """ See :meth:`Loggable._log_init()`. """
+        logger.add_float('Accept.')
+
+    def _log_write(self, logger):
+        """ See :meth:`Loggable._log_write()`. """
+        logger.log(self._mcmc_acceptance)
+
+    def mass(self):
+        """
+        Returns mass ``m`` in kinetic energy function.
+        """
+        return self._m
+
+    def name(self):
+        """ See :meth:`pints.MCMCSampler.name()`. """
+        return 'Monomial-Gamma Hamiltonian Monte Carlo'
+
+    def needs_sensitivities(self):
+        """ See :meth:`pints.MCMCSampler.needs_sensitivities()`. """
+        return True
+
+    def n_hyper_parameters(self):
+        """ See :meth:`TunableMethod.n_hyper_parameters()`. """
+        return 5
+
+    def _pdf(self, p, a, c, m, z):
+        """
+        Auxillary kinetic energy probability density defined in [1].
+        """
+        return (1.0 / z) * np.exp(-self._K_indiv(p, a, c, m))
+
+    def _sample_momentum(self):
+        """
+        Samples from soft kinetic energy pdf defined in [1].
+        """
+        us = np.random.uniform(size=self._n_parameters)
+        return np.array([self._f([u])[0] for u in us])
+
+    def scaled_epsilon(self):
+        """
+        Returns scaled epsilon used in leapfrog algorithm.
+        """
+        return self._scaled_epsilon
+
+    def set_a(self, a):
+        """
+        Sets ``a`` in kinetic energy function.
+        """
+        if a <= 0:
+            raise ValueError("a must be positive")
+        self._a = a
+        self._initialise_ke()
```

Comment:
Are these calls needed? The `self._f` set by this method is only used when running (and when initialise is guaranteed to have been set)